semantics/src/main/kotlin/com/strumenta/kolasu/semantics/symbol/provider/declarative/DeclarativeSymbolProvider.kt

```
@@ -0,0 +1,146 @@
+        @Suppress("UNCHECKED_CAST")
+        return (node as? NodeTy)?.let {
+            this.specification(DeclarativeSymbolProviderRuleArguments(node, symbolProvider))
+            identifierProvider.getIdentifierFor(node)?.let { identifier ->
+                SymbolDescription(identifier, getTypes(node), getProperties(identifierProvider, node))
+            }
+        }
+    }
+
+    fun isCompatibleWith(nodeType: KClass<*>): Boolean {
+        return this.nodeType.isSuperclassOf(nodeType)
+    }
+
+    override fun compareTo(other: DeclarativeSymbolProviderRule<out Node>): Int {
+        return when {
+            this.nodeType.isSuperclassOf(other.nodeType) -> 1
+            other.nodeType.isSuperclassOf(this.nodeType) -> -1
+            else -> (this.nodeType.qualifiedName ?: "") compareTo (other.nodeType.qualifiedName ?: "")
+        }
+    }
+
+    private fun getTypes(node: NodeTy): List<String> {
+        return listOfNotNull(
+            node::class.qualifiedName
+        ) + node::class.allSuperclasses.mapNotNull(KClass<*>::qualifiedName)
+    }
+
+    private fun getProperties(identifierProvider: IdentifierProvider, node: NodeTy): Map<String, ValueDescription> {
+        return this.properties.mapNotNull { (key, value) ->
+            value(identifierProvider, node)?.let { valueDescription -> Pair(key, valueDescription) }
+        }.toMap()
+    }
+
+    private fun toValueDescription(identifierProvider: IdentifierProvider, source: Any?): ValueDescription? {
+        return when (source) {
+            is Boolean -> BooleanValueDescription(source)
+            is Int -> IntegerValueDescription(source)
+            is String -> StringValueDescription(source)
+            is Node -> toContainmentValueDescription(identifierProvider, source)
+            is ReferenceByName<*> -> toReferenceValueDescription(identifierProvider, source)
+            is List<*> -> toListValueDescription(identifierProvider, source)
+            else -> null
```

What about Enum? What about Char? What about arbitrary primitive types? Are they excluded on purpose?

---

Not excluded on purpose, just a matter of time constraints. We can gradually introduce new types while experimenting with the APIs.

---

For values that are currently not supported, should we throw an exception instead of returning null?

---

Okay, I will replace null with a RuntimeException